repo/object/object_manager.go

```diff
@@ -80,6 +80,112 @@ func (om *Manager) Open(ctx context.Context, objectID ID) (Reader, error) {
 	return om.openAndAssertLength(ctx, objectID, -1)
 }
 
+// Concatenate creates an object that's a result of concatenation of other objects. This is more efficient than reading
+// and rewriting the objects because Concatenate can efficiently merge index entries without reading the underlying
+// contents.
+//
+// This function exists primarily to facilitate efficient parallel uploads of very large files (>1GB). Due to bottleneck of
+// splitting which is inherently sequential, we can only one use CPU core for each Writer, which limits throughput.
+//
+// For example when uploading a 100 GB file it is beneficial to independently upload sections of [0..25GB),
+// [25..50GB), [50GB..75GB) and [75GB..100GB) and concatenate them together as this allows us to run four splitters
+// in parallel utilizing more CPU cores. Because some split points now start at fixed bounaries and not content-specific,
+// this causes some slight loss of deduplication at concatenation points (typically 1-2 contents, usually <10MB),
+// so this method should only be used for very large files where this overhead is relatively small.
+func (om *Manager) Concatenate(ctx context.Context, objectIDs []ID) (ID, error) {
+	if len(objectIDs) == 0 {
+		return "", errors.Errorf("empty list of objects")
+	}
+
+	if len(objectIDs) == 1 {
+		return objectIDs[0], nil
+	}
+
+	var (
+		concatenatedEntries []indirectObjectEntry
+		totalLength         int64
+		err                 error
+	)
+
+	for _, objectID := range objectIDs {
+		concatenatedEntries, totalLength, err = om.appendIndexEntriesForObject(ctx, concatenatedEntries, totalLength, objectID)
+		if err != nil {
+			return "", errors.Wrapf(err, "error appending %v", objectID)
+		}
+	}
+
+	log(ctx).Debugf("concatenated: %v total: %v", concatenatedEntries, totalLength)
+
+	w := om.NewWriter(ctx, WriterOptions{
+		Prefix:      indirectContentPrefix,
+		Description: "CONCATENATED INDEX",
+	})
+	defer w.Close() // nolint:errcheck
+
+	if werr := writeIndirectObject(w, concatenatedEntries); werr != nil {
+		return "", werr
+	}
+
+	concatID, err := w.Result()
+	if err != nil {
+		return "", errors.Wrap(err, "error writing concatenated index")
+	}
+
+	return IndirectObjectID(concatID), nil
+}
+
+func (om *Manager) appendIndexEntriesForObject(ctx context.Context, indexEntries []indirectObjectEntry, startingLength int64, objectID ID) (result []indirectObjectEntry, totalLength int64, _ error) {
+	if indexObjectID, ok := objectID.IndexObjectID(); ok {
+		// index object - read its index entries and append to the slice.
+		rd, err := om.Open(ctx, indexObjectID)
+		if err != nil {
+			return nil, 0, errors.Wrapf(err, "error opening index of %v", objectID)
+		}
+		defer rd.Close() //nolint:errcheck
+
+		ndx, err := om.flattenListChunk(rd)
+		if err != nil {
+			return nil, 0, errors.Wrapf(err, "error reading index of %v", objectID)
+		}
```

Comment:
Coincidentally, there's a helper for this in a PR that just got merged. #608

Reply:
thanks, rebased and updated the PR.